Expiring plugin

kylebeee opened the pull request.

joe-p commented on contracts/expiring_plugin.algo.ts:
```diff
@@ -0,0 +1,106 @@
+     * verifyAppCall ensures the external app call being made
+     * is valid and the contract hasn't expired
+     */
+    verify(sender: Account, appCall: AppCallTxn): void {
```
Rather than depending on external method call validation, I feel like it would make more sense to make this a private method and allow apps to inherit from this contract to perform an expiration check.

kylebeee replied:
im trying to think how this would be managed if it wasn't an external contract. If we used inheritance here wouldn't it be difficult to make this something that could be used repetitively? I was thinking you may have users that regularly give a good amount of apps permission to call other methods on your behalf

joe-p replied:
I envision factories being a separate type of application stored in box storage in the abstracted account. Factories would allow for safe creation of plugins with defined functionality for any app. It would basically replace the need for templated apps. For example, if an app wants to do a recurring payment, it would ask you to mint a subscription plugin from the subscription factory with specific parameters. Factories would be called by the user and the factory would have permission to add to the plugins

kylebeee replied:
okay so if were bring factories into the abstracted contract itself we get the benefit of
 - known list of safe or atleast non-malicious factory apps
 - the ability for the factory to add the child contract to the plugin list in an itxn
 
 i wasn't aware we could place contracts in box storage, still piecing this together in my head

kylebeee replied:
ahh i see, so really just extending this from an entirely separate contract to something the contract can manage a list of factories for kinda like we mentioned before about internal factories.

Ill make the adjustments

kylebeee commented:
alright
 - added the `factories` box map
 - added `addFactory`
 - added `removeFactory`
 - added `addFactoryPlugin` A special method for factory plugins to call that ensures the the child app being added to the main plugin list 
 - `rekeyToPlugin` now allows either plugins or factories to do flash rekeys

joe-p commented on contracts/abstracted_account.algo.ts:
```diff
@@ -82,12 +89,12 @@ export class AbstractedAccount extends Contract {
    */
   rekeyToPlugin(plugin: Application): void {
-    assert(this.plugins(plugin).exists);
+    assert(this.plugins(plugin).exists || this.factories(plugin).exists);
```
Under what circumstances would it need to be rekeyed to a factory? I was thinking factories wouldn't require a rekey

kylebeee replied:
my bad, i think i got mixed up here.

The factory would need to use the flash rekey functionality in order to work would it not?

This definitely wont work, because it would give anyone permission to call those factory methods but it still needs to rekey to let the factory add the child plugin to your account right? Just triggered by the EOA or Admin?

joe-p replied:
1. Admin adds factory app to abstracted account. This adds the factory app the abstracted account so it allows plugins to be added from it
1.  Admin calls factory app. This adds the created app to the plugin list

I suppose the key here is that the factory app has a check to ensure the caller is the admin of the abstracted account. 

Now that I think about it though there really isn't much of a benefit from separating factories from plugin. The security assumptions are basically the same because the factory could mint a malicious plugin and then use that plugin.

kylebeee replied:
yeah, i was thinking this through and started to question what the point of the factories box was if the admin is required to call the factory in the first place

joe-p commented on contracts/expiring_plugin.algo.ts:
```diff
@@ -0,0 +1,121 @@
+        )
+
+        // ensure we're calling the correct app from the correct account
+        verifyAppCallTxn(appCall, {
```
Rather than verifying the app call I think we would need to rekey to the app here. Otherwise we need to trust the the app call is only callable when `verify` is called
